teps/0007-conditions-beta.md

```diff
@@ -297,56 +299,88 @@ We can explore adding more [Operators](https://github.com/kubernetes/kubernetes/
-- can take the values `skipBranch` or `runBranch` only; there will be a validation error if anything else is passed to `whenSkipped`
-- can be specified guarded `Tasks` with ordering dependencies only; there will be a validation error if `whenSkipped` is specified in `Tasks` with resource dependencies
+We already support specifying a list of `WhenExpressions` through the `when` field, as described in [Efficiency](#efficiency), with the following syntax:
+
+```yaml
+when:
+  - input: 'foo'
+    operator: in
+    values: ['bar']
+```
+
+To provide more flexibility when `WhenExpressions` evaluate to `False`, we propose adding `expressions` and `scope` fields under the `when` field where:
+- The `expressions` field will contain the list of `WhenExpressions` as described in [Efficiency](#efficiency)
+- The `scope` fields will contain the scope of the `WhenExpressions`, that is whether the `WhenExpressions` guard the `Node` only or the whole `Branch`
 
-Here's how a user can use `whenSkipped` to execute ordering-dependent tasks:
+The syntax would be:
 
 ```yaml
-api: tekton.dev/v1beta1
-kind: Pipeline
-metadata:
-    name: generate-file
-spec:
-  workspaces:
-    - name: source-repo
-  params:
-    - name: path
-      default: 'README.md'
-  tasks:
-    - name: create-file # executed
-      taskRef:
-        name: create-readme-file
-      workspaces:
-        - name: source
-          workspace: source-repo
-    - name: file-exists # executed
-      taskRef:
-        name: file-exists
-      workspaces:
-        - name: source
-          workspace: source-repo
-    - name: echo-file-does-not-exist # skipped
-      when:
+when:
+  scope: Node/Branch
+  expressions:
+    - input: 'foo'
+      operator: in
+      values: ['bar']
+```
+
+The `scope` field is optional, defaults to `Branch` and can be set to `Branch` to specify that the `Node`'s `WhenExpressions` guard the execution of the `Branch` (the `Node` and its dependent `Nodes`). However, when `Node` has ordering dependencies only, users can set the `scope` field to `Node` to specify that the `Node`'s `WhenExpressions` guard the execution of the `Node` only, thus the all dependent nodes (which are ordering dependent tasks only) can be executed. 
+
+Note that the `scope` field:
+- can take the values `Branch` or `Node` only; there will be a validation error if anything else is passed to `scope`
+- can take the value `Node` in guarded `Nodes` with ordering dependencies only; there will be a validation error if `scope` is set to `Node` in `Nodes` with resource dependencies
+
+The latter syntax allows us to make it explicit that the skipping policy, as specified by `WhenScope`, is related to the `WhenExpressions`. We can support both the former and latter syntaxes initially then give users 9 months to migrate to the new syntax, per [the beta api policy](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md). 
```

Comment:
I am interested to know a bit more about supporting both the existing array syntax and the new syntax.  I couldn't think of an example where I'd seen this before.  The only type like this in our code I'm aware of are our params, where the user can specify a `value` of either an array or a string.  However supporting that type has also been a source of [one specific issue that we don't have a great fix for right now](https://github.com/tektoncd/pipeline/issues/3323).

I'm not sure how important that specific issue is to the pipeline project but I thought I would flag it up.  Wondering if there's another way to approach this kind of "union" type in our API without hitting that same problem.

Reply:
thanks for bringing this up @sbwsg!

yes, we'd support both syntaxes in a way that's similar to how we do it in `ArrayOrStrings` in `Params`

but while that is the "permanent" state in `Params`, it would be a temporary fix in `WhenExpressions` for 9 months only to give users enough time to migrate to the new syntax then we can stop supporting the old syntax

i think temporarily supporting both syntaxes to give users a better experience transitioning outweighs the issue of temporarily confusing tools (also waiting for more details from the contributor to further gauge the severity), what do you think?

Reply:
Ah, good point, I hadn't factored in that this will be a temporary bridge from one to the other.  Makes sense!